app/services/catalog/update_order_item.rb

```diff
@@ -0,0 +1,43 @@
+module Catalog
+  class UpdateOrderItem
+    class OrderItemNotFound < StandardError; end
+
+    def initialize(topic)
+      @payload = topic.payload
+      @message = topic.message
+    end
+
+    def process
+      Rails.logger.info("Processing service order topic message: #{@message} with payload: #{@payload}")
+
+      Rails.logger.info("Searching for OrderItem with a task_ref: #{@payload["task_id"]}")
+      order_item = find_order_item
+      Rails.logger.info("Found OrderItem: #{order_item.id}")
+
+      ManageIQ::API::Common::Request.with_request(order_item.context.transform_keys(&:to_sym)) do
+        order_item.update_message("info", "Task update message received with payload: #{@payload}")
+
+        if @payload["state"] == "completed"
+          order_item.state = "Order Completed"
+          order_item.update_message("info", "Order Complete")
+
+          Rails.logger.info("Updating OrderItem: #{order_item.id} with 'Order Completed' state")
+          order_item.save!
+          Rails.logger.info("Finished updating OrderItem: #{order_item.id} with 'Order Completed' state")
```

Comment:
@eclarizio Somewhere around here we have to call this method to mark the order as finalized.
https://github.com/ManageIQ/catalog-api/blob/6123893880c14260bd9571cae9caf044bc1cedaa/app/models/order.rb#L19

And some of this code also needs to be moved or deleted'
https://github.com/ManageIQ/catalog-api/blob/6123893880c14260bd9571cae9caf044bc1cedaa/app/models/order_item.rb#L57

https://github.com/ManageIQ/catalog-api/blob/6123893880c14260bd9571cae9caf044bc1cedaa/app/models/order_item.rb#L68

https://github.com/ManageIQ/catalog-api/blob/6123893880c14260bd9571cae9caf044bc1cedaa/app/models/order_item.rb#L74

Reply:
This feels like something that should be added (and the other code moved/deleted) in a separate PR, since it's just adding scope to this PR, and should be a simpler change.

Reply:
@eclarizio You can do that in a separate PR, currently there isn't anything that is marking the order as completed.